hooks/boot/20-certs_update.rb

```diff
@@ -0,0 +1,36 @@
+# Add options around regenerating certificates
+if module_enabled?('katello')
```

Comment:
Is this really correct? I think these cert options are also relevant in other cases, like content proxies. Isn't it sufficient to check if the certs module is enabled?
```suggestion
if module_enabled?('certs')
```

(repeated throughout this PR)

However, a potentially larger issue is access to variables in the boot stage. https://github.com/theforeman/kafo#hooks isn't explicit about this. IMHO these methods should raise exceptions when used without initialization.

Reply:
What does used without initialization here mean?

Reply:
Thinking more about it, technically the boot hook shouldn't check for enabled but rather whether the module is present in the answers file. This is because answers are not yet determined when booting.

You can run `--no-enable-certs` (which would save the answers) and in the next run `--enable-certs --certs-update-server`. With the current approach that would fail because the option `--certs-update-server` is not added, but it's a valid flow. Probably more theoretical, but let's use this as a learning experience about how things are supposed to work.

[This](https://github.com/theforeman/kafo/blob/68ae1ad760f4073a6fedc9ce8cfe4f6ddd1a85e8/lib/kafo/hook_context.rb#L60-L66) is the source for `module_enabled?`. With that, it's easy to see how `module_present?(module_name)` would be implemented.

Again, not sure if at this stage it's really present. Easy to verify though:
```console
bundle update
rm -f Puppetfile.lock # TODO: might be nice if rake clean removed this when it's in .gitignore
bundle exec rake clean build
bundle exec ./bin/foreman-installer --scenario katello --help
```

Note it doesn't work no my machine because it fails on getting defaults (`This module does not support osfamily Gentoo`) but on Fedora it worked last time I checked.

Reply:
> What does used without initialization here mean?

During the boot phase (as you can see in the linked docs) kafo is not initialized. It means you don't have a logger (yet) and parameters have no values. Perhaps the modules are loaded since the configs are loaded but I'd really verify this.

Reply:
During `bundle exec rake clean build`,

```
W, [2020-07-13T13:57:12.530656 #2304132]  WARN -- : Parser cache ./_build/parser_cache/katello.yaml is from a different version of Kafo, skipping setup
W, [2020-07-13T13:57:13.552531 #2304132]  WARN -- : Parser cache ./_build/parser_cache/katello.yaml is from a different version of Kafo, skipping setup
W, [2020-07-13T13:57:14.595790 #2304132]  WARN -- : Parser cache ./_build/parser_cache/katello.yaml is from a different version of Kafo, skipping setup
```

And later,

```
W, [2020-07-13T13:58:12.694403 #2304347]  WARN -- : Parser cache ./_build/parser_cache/foreman-proxy-certs.yaml is from a different version of Kafo, skipping setup
W, [2020-07-13T13:58:13.721999 #2304347]  WARN -- : Parser cache ./_build/parser_cache/foreman-proxy-certs.yaml is from a different version of Kafo, skipping setup
```
Then, during `bundle exec ./bin/foreman-installer --scenario katello --help`:

`Could not get default values, check log file at ./_build//katello.log for more information`

Checking the log,

`[ERROR 2020-07-13T13:58:35 main] Could not get default values, cannot continue`

The Kafo version,

```
$ bundle info kafo
  * kafo (4.1.0)
	Summary: A gem for making installations based on puppet user friendly
	Homepage: https://github.com/theforeman/kafo
	Path: /usr/share/gems/gems/kafo-4.1.0
```

But I am unsure how to check what version was used with `./_build/parser_cache/{katellomforeman-proxy-certs}.yaml`

Reply:
Did you check `./_build//katello.log`?

Reply:
It turns out the issue is

```
$ grep Function _build/katello.log
[ERROR 2020-07-14T10:54:13 main] Error: Evaluation Error: Error while evaluating a Function Call, 'upcase' parameter 'arg' expects a value of type Numeric, String, Array, or Hash, got Undef (file: /home/wclark/Projects/foreman-installer/_build/modules/foreman_proxy/manifests/params.pp, line: 278, column: 29) on node prometheus

$ sed -n '277,278p;279q' /home/wclark/Projects/foreman-installer/_build/modules/foreman_proxy/manifests/params.pp
  $dns_zone               = $facts['networking']['domain']
  $dns_realm              = upcase($dns_zone)

$ facter networking.dns.domain

$
```

I'm not sure how exactly facter determines this fact but it's not too surprising since I previously only assigned a shortname to this machine. I will get that sorted and try again.

Reply:
Yes, that's actually a bug that we should fix. Various modules we maintain don't gracefully fall back if there's no domain. Even if variables end up being unused (like realm being turned off) they hard fail. Code like this is actually what we should be used.

https://github.com/theforeman/puppet-dhcp/blob/5ba9415784e47c4e8a65f51a679d721568e82b3b/manifests/params.pp#L5-L9

Then when we know the variable is used, we should have checks.

Reply:
This can now use `module_present?` helper